File: sdk/elixir/lib/rocksky/dropbox.ex
defmodule Rocksky.Dropbox do
  @moduledoc "`app.rocksky.dropbox.*` endpoints. Require an authenticated client."

  alias Rocksky.HTTP

  @doc "List files at a Dropbox path. Params: `:at`."
  def get_files(client, params \\ []),
    do: HTTP.query(client, "app.rocksky.dropbox.getFiles", params)

  @doc "Metadata for a Dropbox path. Params: `:path`."
  def get_metadata(client, params \\ []),
    do: HTTP.query(client, "app.rocksky.dropbox.getMetadata", params)

  @doc "Generate a temporary download link. Params: `:path`."
  def get_temporary_link(client, params \\ []),
    do: HTTP.query(client, "app.rocksky.dropbox.getTemporaryLink", params)

  @doc "Download a file by Dropbox file id. Params: `:fileId`."
  def download_file(client, params \\ []),
    do: HTTP.query(client, "app.rocksky.dropbox.downloadFile", params)
end

File: sdk/clojure/src/rocksky/dropbox.clj
(ns rocksky.dropbox
  "Endpoints under app.rocksky.dropbox.*"
  (:require [rocksky.client :as c]))

(defn get-files
  "List files in Dropbox.

  Optional: `:at` (folder path)."
  ([client] (get-files client nil))
  ([client {:keys [at]}]
   (c/query client :app.rocksky.dropbox.getFiles {:at at})))

(defn get-metadata
  "Get metadata of a file or folder. Required: `:path`."
  [client {:keys [path]}]
  (c/query client :app.rocksky.dropbox.getMetadata {:path path}))

(defn get-temporary-link
  "Get a temporary link to a file. Required: `:path`."
  [client {:keys [path]}]
  (c/query client :app.rocksky.dropbox.getTemporaryLink {:path path}))

(defn download-file
  "Download a file by `:file-id`. Returns raw bytes — wrap the URL yourself
  if you'd rather stream. Required: `:file-id`."
  [client {:keys [file-id]}]
  (c/query client :app.rocksky.dropbox.downloadFile {:fileId file-id}))
